src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,575 @@
+            }
+            return new KDF((KDFSpi) instance.impl, instance.provider, algorithm,
+                           algParameterSpec);
+
+        } catch (NoSuchAlgorithmException nsae) {
+            return handleException(nsae);
+        }
+    }
+
+    /**
+     * Creates an instance of the {@code KDF} object using a supplied
+     * {@code Provider} instance.
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use
+     * @param algParameterSpec
+     *     the {@code AlgorithmParameterSpec} used to configure this KDF's
+     *     algorithm or {@code null} if no additional parameters were provided.
+     * @param provider
+     *     the provider
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if no {@code Provider} supports a {@code KDFSpi} implementation for
+     *     the specified algorithm
+     * @throws InvalidAlgorithmParameterException
+     *     if the {@code AlgorithmParameterSpec} is an invalid value
+     * @throws NullPointerException
+     *     if the algorithm is {@code null}
+     */
+    public static KDF getInstance(String algorithm,
+                                  AlgorithmParameterSpec algParameterSpec,
+                                  Provider provider)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Objects.requireNonNull(algorithm, "null algorithm name");
+        try {
+            Instance instance = GetInstance.getInstance("KDF", KDFSpi.class,
+                                                        algorithm,
+                                                        algParameterSpec,
+                                                        provider);
+            if (!JceSecurity.canUseProvider(instance.provider)) {
+                String msg = "JCE cannot authenticate the provider "
+                             + instance.provider.getName();
+                throw new SecurityException(msg);
+            }
+            return new KDF((KDFSpi) instance.impl, instance.provider, algorithm,
+                           algParameterSpec);
+
+        } catch (NoSuchAlgorithmException nsae) {
+            return handleException(nsae);
+        }
+    }
+
+    private static KDF handleException(NoSuchAlgorithmException e)
+        throws NoSuchAlgorithmException,
+               InvalidAlgorithmParameterException {
+        Throwable cause = e.getCause();
+        if (cause instanceof InvalidAlgorithmParameterException) {
+            throw (InvalidAlgorithmParameterException) cause;
+        }
+        throw e;
+    }
+
+    /**
+     * Derive a key, returned as a {@code SecretKey}.
+     * <p>
+     * The {@code deriveKey} method may be called multiple times once a
+     * {@code KDF} object is initialized.
+     * <p>
+     * Delayed provider selection is also supported such that the provider
+     * performing the derive is not selected until the method is called.
```

Comment:
Delayed provider selection is an important enough topic that it probably should be in the class summary. However it is complicated to word correctly as there is also the case if someone calls `getProviderName` beforehand which locks the provider to the first one supporting the algorithm. I would probably also avoid "delayed provider" as that is not a term currently used in the javadocs. Suggest something like:

If a provider is not specified in the getInstance method when instantiating a KDF object, the provider is selected the first time the deriveKey or deriveData method is called and a provider is chosen that supports the parameters passed to the deriveKey or deriveData method, for example the initial key material. However, if getProviderName is called before calling the deriveKey or deriveData methods, the first provider supporting the KDF algorithm is chosen which may not be the desired one; therefore it is recommended to not call getProviderName until after a key derivation operation.

Reply:
This is because the selection occurs just once. Should we explicitly mention this?

Reply:
Yes, maybe add at end: "Once a provider is selected, it cannot be changed."